👷 Add client document (#159)
gofiber/docs: api/client.md

---
description: >-

>   The Client struct represents the HTTP Client. It has methods like Get, Post, etc, they'll create a new Agent instance and parse request url automatically.

Reviewer: ```suggestion
  The Client struct represents the Fiber HTTP Client.
  
  
```

---

# ⚡ Client

## Start request
Start a http request with http method and url.

{% code title="Signatures" %}
```go
// Client http methods
func (c *Client) Get(url string) *Agent
func (c *Client) Head(url string) *Agent
func (c *Client) Post(url string) *Agent
func (c *Client) Put(url string) *Agent
func (c *Client) Patch(url string) *Agent
func (c *Client) Delete(url string) *Agent
```
{% endcode %}

## ✨ Agent

> `Agent` is on top of [fasthttp `HostClient`](https://github.com/valyala/fasthttp/blob/master/client.go#L603) which has lots of convenient helper methods and several `send request` methods.

Reviewer: ```suggestion
`Agent` is built on top of FastHTTP's [`HostClient`](https://github.com/valyala/fasthttp/blob/master/client.go#L603) which has lots of convenient helper methods and several `send request` methods.
```

> `Agent` is on top of [fasthttp `HostClient`](https://github.com/valyala/fasthttp/blob/master/client.go#L603) which has lots of convenient helper methods and several `send request` methods.

Reviewer: By send request methods do you mean like GET & POST?

### Parse

> Parse initializes URI and HostClient.

Reviewer: Sounds like Parse takes a URI as a param

```suggestion
Parse initializes a HostClient.
```

Author: > Sounds like Parse takes a URI as a param

Parse parse the total request, users should set URI via agent.Request().

{% code title="Parse" %}
```go
a := AcquireAgent()
req := a.Request()
req.Header.SetMethod(MethodGet)
req.SetRequestURI("http://example.com")

if err := a.Parse(); err != nil {
	panic(err)
}

code, body, errs := a.Bytes() // ...
```
{% endcode %}

### Set

> Set sets the given 'key: value' header.

Reviewer: ```suggestion
Set sets the given `key: value` header.
```

{% code title="Signature" %}
```go
func (a *Agent) Set(k, v string) *Agent
func (a *Agent) SetBytesK(k []byte, v string) *Agent
func (a *Agent) SetBytesV(k string, v []byte) *Agent
func (a *Agent) SetBytesKV(k []byte, v []byte) *Agent
```
{% endcode %}

{% code title="Example" %}
```go

agent.Set("k1", "v1").
    SetBytesK([]byte("k1"), "v1").
    SetBytesV("k1", []byte("v1")).
    SetBytesKV([]byte("k2"), []byte("v2"))
// ...
```
{% endcode %}

### Add

> Add adds the given 'key: value' header. Multiple headers with the same key may be added with this function.

Reviewer: ```suggestion
Add adds the given `key: value` header. Multiple headers with the same key may be added with this function.
```

> Add adds the given 'key: value' header. Multiple headers with the same key may be added with this function.

Reviewer: > Multiple headers with the same key may be added with this function.
So does this mean the header's won't overwrite eachother?

Author: Yeah you are right. we can have headers like
```
K1: v1
K1: v2
```

{% code title="Signature" %}
```go
func (a *Agent) Add(k, v string) *Agent
func (a *Agent) AddBytesK(k []byte, v string) *Agent
func (a *Agent) AddBytesV(k string, v []byte) *Agent
func (a *Agent) AddBytesKV(k []byte, v []byte) *Agent
```
{% endcode %}

{% code title="Example" %}
```go

agent.Add("k1", "v1").
    AddBytesK([]byte("k1"), "v1").
    AddBytesV("k1", []byte("v1")).
    AddBytesKV([]byte("k2"), []byte("v2"))

> // ...

Reviewer: We should show what the headers look like:

```suggestion
// Headers:
// k1: v1
// k1: v1
// k2: v2
```

```
{% endcode %}

### ConnectionClose

> ConnectionClose sets 'Connection: close' header.

Reviewer: ```suggestion
ConnectionClose adds the `Connection: close` header.
```

{% code title="Signature" %}
```go
func (a *Agent) ConnectionClose() *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.ConnectionClose()
// ...
```
{% endcode %}

### UserAgent

> UserAgent sets User-Agent header value.

Reviewer: ```suggestion
UserAgent sets `User-Agent` header value.
```

{% code title="Signature" %}
```go
func (a *Agent) UserAgent(userAgent string) *Agent
func (a *Agent) UserAgentBytes(userAgent []byte) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.UserAgent("fiber")
// ...
```
{% endcode %}

### Cookie

> Cookie sets 'key: value' cookie.

Reviewer: ```suggestion
Cookie sets a cookie in `key: value` form. `Cookies` can be used to set multiple cookies.
```

{% code title="Signature" %}
```go
func (a *Agent) Cookie(key, value string) *Agent
func (a *Agent) CookieBytesK(key []byte, value string) *Agent
func (a *Agent) CookieBytesKV(key, value []byte) *Agent
func (a *Agent) Cookies(kv ...string) *Agent
func (a *Agent) CookiesBytesKV(kv ...[]byte) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.Cookie("k", "v")

> agent.Cookies("k1", "v1", "k2", "v2")

Reviewer: Are we able use a map too?

```
m := map[string]string{
  "k1": "v1",
  "k2": "v2",
}
agent.Cookies(m)
```

Author: Yes we can. But if we do so, we have to declare a map struct.

// ...
```
{% endcode %}

### Referer

> Referer sets Referer header value.

Reviewer: ```suggestion
Referer sets the Referer header value.
```

{% code title="Signature" %}
```go
func (a *Agent) Referer(referer string) *Agent
func (a *Agent) RefererBytes(referer []byte) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.Referer("https://docs.gofiber.io")
// ...
```
{% endcode %}

### ContentType
ContentType sets Content-Type header value.

{% code title="Signature" %}
```go
func (a *Agent) ContentType(contentType string) *Agent
func (a *Agent) ContentTypeBytes(contentType []byte) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.ContentType("custom-type")
// ...
```
{% endcode %}

### Host

> Host sets host for the uri.

Reviewer: ```suggestion
Host sets the Host header.
```

{% code title="Signature" %}
```go
func (a *Agent) Host(host string) *Agent
func (a *Agent) HostBytes(host []byte) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.Host("example.com")
// ...
```
{% endcode %}

### QueryString

> QueryString sets URI query string.

Reviewer: ```suggestion
QueryString sets the URI query string.
```

{% code title="Signature" %}
```go
func (a *Agent) QueryString(queryString string) *Agent
func (a *Agent) QueryStringBytes(queryString []byte) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.QueryString("foo=bar")
// ...
```
{% endcode %}

### BasicAuth

> BasicAuth sets URI username and password.

Reviewer: ```suggestion
BasicAuth sets the URI username and password using HTTP Basic Auth.
```

{% code title="Signature" %}
```go
func (a *Agent) BasicAuth(username, password string) *Agent
func (a *Agent) BasicAuthBytes(username, password []byte) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.BasicAuth("foo", "bar")
// ...
```
{% endcode %}

### Body
There are several ways to set request body.

{% code title="Signature" %}
```go
func (a *Agent) BodyString(bodyString string) *Agent
func (a *Agent) Body(body []byte) *Agent

// BodyStream sets request body stream and, optionally body size.
//
// If bodySize is >= 0, then the bodyStream must provide exactly bodySize bytes
// before returning io.EOF.
//
// If bodySize < 0, then bodyStream is read until io.EOF.
//
// bodyStream.Close() is called after finishing reading all body data
// if it implements io.Closer.
//
// Note that GET and HEAD requests cannot have body.
func (a *Agent) BodyStream(bodyStream io.Reader, bodySize int) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.BodyString("foo=bar")
agent.Body([]byte("bar=baz"))
agent.BodyStream(strings.NewReader("body=stream"), -1)
// ...
```
{% endcode %}

### JSON

> JSON sends a JSON request with `application/json` content type.

Reviewer: ```suggestion
JSON sends a JSON request by setting the Content-Type header to `application/json`.
```

{% code title="Signature" %}
```go
func (a *Agent) JSON(v interface{}) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.JSON(fiber.Map{"success": true})
// ...
```
{% endcode %}

### XML

> XML sends an XML request with `application/xml` content type.

Reviewer: ```suggestion
XML sends an XML request by setting the Content-Type header to `application/xml`.
```

{% code title="Signature" %}
```go
func (a *Agent) XML(v interface{}) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.XML(fiber.Map{"success": true})
// ...
```
{% endcode %}

### Form

> Form sends form request with `application/x-www-form-urlencoded` content type.

Reviewer: ```suggestion
Form sends a form request by setting the Content-Type header to `application/x-www-form-urlencoded`.
```

{% code title="Signature" %}
```go
// Form sends form request with body if args is non-nil.
//
// It is recommended obtaining args via AcquireArgs and release it
// manually in performance-critical code.
func (a *Agent) Form(args *Args) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
args := AcquireArgs()
args.Set("foo", "bar")

agent.Form(args)
// ...
ReleaseArgs(args)
```
{% endcode %}

### MultipartForm

> MultipartForm sends multipart form request with `multipart/form-data` content type, k-v and files.

Reviewer: ```suggestion
MultipartForm sends multipart form request by setting the Content-Type header to `multipart/form-data`. These requests can include key-value's and files.
```

{% code title="Signature" %}
```go
// MultipartForm sends multipart form request with k-v and files.
//

> // It is recommended obtaining args via AcquireArgs and release it

Reviewer: ```suggestion
// It is recommended to obtain args via AcquireArgs and release it
```

// manually in performance-critical code.
func (a *Agent) Form(args *Args) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
args := AcquireArgs()
args.Set("foo", "bar")

agent.MultipartForm(args)
// ...
ReleaseArgs(args)
```
{% endcode %}

Fiber provides several methods for sending files. Note that they must be called before `MultipartForm`. 

#### Boundary
Boundary sets boundary for multipart form request.

{% code title="Signature" %}
```go
func (a *Agent) Boundary(boundary string) *Agent
```
{% endcode %}

{% code title="Example" %}
```go
agent.Boundary("myBoundary")
    .MultipartForm(nil)
// ...
```
{% endcode %}

#### SendFile(s)

> SendFile(s) read file(s) and appends it(them) to multipart form request.

Reviewer: ```suggestion
SendFile read a file and appends it to a multipart form request. Sendfiles can be used to append multiple files.
```
